### ISSUE TITLE
VSCode: ERB file false positive problems

### ISSUE THREAD

**Issue by bananastalktome** [Shopify/ruby-lsp#2282]
### Description
The latest Ruby LSP VSCode extension (v0.7.11) generates numerous code problems for ERB files which should not be indicated as problems. The issue does not occur with v0.7.10. I've attached two screenshots - one of the code:
![code](https://github.com/Shopify/ruby-lsp/assets/60870/c9156d1b-1da6-4e29-81e8-f0c32f96ef2c)

and another of the 'Problems' output:
![problems](https://github.com/Shopify/ruby-lsp/assets/60870/f8fdefe5-3a55-4740-80ec-7eb41cc7d539)

I think I narrowed it down to it only occurring when the `solargraph` gem is installed from my `Gemfile`, but the odd behavior occurs whether the VSCode "castwide.solargraph" extension is enabled or not (even occurs with the `solargraph` gem present when the "castwide.solargraph" extension has been fully uninstalled).

### To Reproduce
1. Create a new empty directory with the following files:
   * `.ruby-version` containing an rbenv ruby version (3.2.2 in my case)
   * `Gemfile` containing only `gem "solargraph", require: false`
   * A dummy `test.html.erb` file with any valid HTML
2. run `bundle install` and then open VSCode in that directory
3. Notice the problems in the `test.html.erb` file once the LSP extension starts the language server

**Comment by remy727** [comment]
@vinistock, thank you for your suggestion. 👍 
Yes, downgrading the extension to `v0.7.10` works.

**Comment by jonknapp** [comment]
Found this issue after running into the same issue with erb file. Reverting to an older version worked. The file is listed as `erb` and my settings don't have any associations for `*.erb`

<details><summary>settings</summary>

```json
{
  "[javascript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[markdown]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true
  },
  "[yaml]": {},
  "deno.enable": false,
  "workbench.startupEditor": "newUntitledFile",
  "files.trimTrailingWhitespace": true,
  "editor.tabSize": 2,
  "files.insertFinalNewline": true,
  "extensions.ignoreRecommendations": false,
  "gitlens.currentLine.enabled": false,
  "gitlens.hovers.currentLine.over": "line",
  "gitlens.codeLens.enabled": false,
  "editor.rulers": [80, 120],
  "terraform.indexing": {
    "enabled": false,
    "liveIndexing": false,
    "delay": 500,
    "exclude": [".terraform/**/*", "**/.terraform/**/*"]
  },
  "gitlens.advanced.messages": {
    "suppressLineUncommittedWarning": true
  },
  "[rust]": {
    "editor.defaultFormatter": "matklad.rust-analyzer"
  },
  "editor.formatOnSave": true,
  "editor.renderWhitespace": "boundary",
  "rust-client.disableRustup": true,
  "search.useGlobalIgnoreFiles": true,
  "[json]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "javascript.updateImportsOnFileMove.enabled": "always",
  "rust-analyzer.diagnostics.disabled": ["missing-unsafe"],
  "[javascriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[jsonc]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "workbench.editorAssociations": {
    "*.ipynb": "jupyter.notebook.ipynb"
  },
  "[sql]": {
    "editor.defaultFormatter": "adpyke.vscode-sql-formatter"
  },
  "window.restoreWindows": "none",
  "[nix]": {
    "editor.defaultFormatter": "jnoortheen.nix-ide"
  },
  "nix.serverPath": "nil",
  "files.associations": {
    "COMMIT_EDITMSG": "git-commit"
  },
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[html]": {
    "editor.defaultFormatter": "vscode.html-language-features"
  },
  "files.maxMemoryForLargeFilesMB": 8192,
  "terraform.languageServer.enable": true,
  "redhat.telemetry.enabled": false,
  "nix.serverSettings": {},
  "workbench.editor.empty.hint": "hidden",
  "[typescriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "rubyLsp.rubyVersionManager": {
    "identifier": "none"
  },
  "window.zoomLevel": 1,
  "git.enableCommitSigning": true,
  "rubyLsp.enableExperimentalFeatures": true,
  "[css]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[scss]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  }
}
```
</details>

**Comment by bananastalktome** [comment]
@vinistock I did have `craigmaslowski.erb` and `aliariff.vscode-erb-beautify`, but even after disabling them & restarting VSCode completely, I still get the highlighting issue with `0.7.11` (though now my language shows `erb` just like in https://github.com/Shopify/ruby-lsp/issues/2282#issuecomment-2221211980). I don't define that `files.associations` `*.erb` key anywhere.

**Comment by vinistock** [comment]
If you turn on verbose tracing for the server:

```jsonc
// Settings.json
{
  "ruby lsp.trace.server": "verbose"
}
```

All requests and their parameters get printed to the `Ruby LSP` channel in the `Output` tab. When you open an ERB file, the `textDocument/didOpen` request should have the `languageId` parameter set to `erb`.

If that's not the case, something is changing the language ID of ERB files. It could be a user setting, a workspace setting (.vscode/setings.json), a different extension. We need to understand what is changing the language ID.

**Comment by bananastalktome** [comment]
```
2024-07-10 15:33:44.818 [info] (project) [Trace - 3:33:44 PM] Sending notification 'textDocument/didOpen'.
2024-07-10 15:33:44.818 [info] (project) Params: {
    "textDocument": {
        "uri": "file:///Users/<REDACTED>file.html.erb",
        "languageId": "erb",
        "version": 1,
        "text": "<REDACTED>"
    }
}
```

**Comment by vinistock** [comment]
So, the language ID is correct. Oh, maybe the server is outdated? If you do

1. CMD + SHIFT + P
2. Update language server gem

Does that fix the issue?

**Comment by bananastalktome** [comment]
@vinistock I ran that server gem update VSCode command, but that didn't fix the issue either, unfortunately.

**Comment by vinistock** [comment]
If you click on the language status item (the `{}` next to Ruby [here](https://github.com/Shopify/ruby-lsp/tree/main/vscode#configuration)), what server version are you on?

**Comment by vinistock** [comment]
Okay. That makes sense, the server isn't able to handle ERB files.

Usually, when it fails to upgrade, it's because your project is locked to an older version of `prism` or `sorbet-runtime`, which then stops Bundler from updating the LSP.

You can try it on the terminal with

```shell
BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle update ruby-lsp
```

**Comment by bananastalktome** [comment]
humm, `Bundler attempted to update ruby-lsp but its version stayed the same`. I had to manually add `gem "ruby-lsp", "0.17.6"` to my project `Gemfile` (not ideal, but it worked) in order to update to the latest version, and then erb mode works. I'd rather not pin the ruby-lsp version in my Gemfile, it's weird that it wouldn't update when I ran the command even in the basic Gemfile reproduction setup (with only `solargraph` listed).

**Comment by vinistock** [comment]
It could be the `rbs` dependency that is restricting it as well. Solargraph also depends on RBS, so if it's locked to anything older than 3, the Ruby LSP won't update.

**Comment by LucasKuhn** [comment]
Thank you @vinistock! Deleted my suggestion to avoid confusion. 
For me it was indeed `rbs`, required by solargraph.

**Comment by Syntaf** [comment]
Ran into this issue tonight,I noticed I still had an old version of solargraph listed in my development group despite not using that extension anymore:

```
  gem 'solargraph', '~> 0.49.0'
```

Removing this gem worked -- not a great solution if you use solargraph but if you don't and have the gem just remove it to resolve the issue.

**Comment by earnold** [comment]
Removing `solargraph` also solved this for me.

**Comment by nikasvan** [comment]
Thanks for suggestion @vinistock Downgrading to `v0.7.10` worked for me. Hope this gets fixed 🤞

**Comment by Earlopain** [comment]
What are your reasons for keeping solargraph around still? With https://gist.github.com/castwide/28b349566a223dfb439a337aea29713e and the `solargraph-rails` extension it does still offer better rails support (I think), just has better insight into the metaprogramming. I'm curious if that is part of your reasoning, or if its something else. The fix from the solargraph side would be https://github.com/castwide/solargraph/pull/662.

**Comment by hayesr** [comment]
FWIW, my problem was that I'm getting ruby-lsp via the ruby-lsp-rails gem (ruby-lsp-rails is the only thing my Gemfile), so trying to `bundle update ruby-lsp` was not getting me to the latest version. Had to `bundle update ruby-lsp-rails` to unjam the bundles.

**Comment by hayesr** [comment]
@andyw8 Ok that's helpful, thanks!

**Comment by vinistock** [comment]
@nikasvan I don't believe downgrading is an optimal solution. I suggested that while we were still understanding the issue. But it goes beyond not having errors for ERB files.

If you are unable to upgrade the server gem, you will not have access to any new features we ship. Almost all features in a language server are implemented in the server and not on the client side.

For example, if you can't upgrade past v0.17.2, you are currently missing out on:

- Support for method aliases
- Support for Ruby core declarations through RBS
- Proper constant resolution (thanks to the work for ancestor linearization)
- Singleton context handling, which translates to go to definition, completion, hover and signature help for singleton methods, class instance variables and constants defined inside a singleton block
- Quickfixes for contextual RuboCop violations
- The super type hierarchy visualization feature
- ERB features
- Completion for locals
- Support for `super` in go to definition and hover

Plus whatever we ship in the next versions until the constraint is lifted and you can successfully upgrade the `ruby-lsp` gem.

Even if we came up with a way to avoid the errors for ERB, being unable to upgrade the language server gem is incompatible with delivering a good developer experience. We need to be able to get new features into the hands of developers. Bugfixes. Performance improvements. And the quicker we can do that, the better.

**Comment by nimmolo** [comment]
Thanks for digging into this @vinistock and everybody else.

For our project, it's another gem `prawn-manual_builder` that is locking the version of `prism` at `0.22.0`. I'll raise the issue with them.

**Comment by feliperaul** [comment]
I was having the same issue and, after reading this issue start to finish, these were the steps:

- Removing `solargraph` and `ruby-lsp-rails` from gemfile;
- Running `bundle`;
- Uninstalling Ruby Solargraph extension;
- Activating the command palette and running `Ruby LSP: update language server gem`;
- Clicking on the `{ }` icon in the bottom right corner of vscode and confirming Ruby LSP server is version >= 0.17.7

This solved the ERB issue and the code navigation kept working!

Thanks a lot for the work on this great DX tool.

**Comment by andyw8** [comment]
> Alternatively, we could also just print a warning linking to this issue and explaining what's wrong, but not automatically restart the client - but then the ERB issues will not go away.

This seems preferable to me, since it gives more of an opportunity to explain the problem to users.

**Comment by vinistock** [comment]
I really disliked the double-launch approach, so we ended up providing the ability to disable ERB support through a setting https://github.com/Shopify/ruby-lsp/pull/2311.

## For users who find this issue in the future:

### I'm having trouble with ERB files

The reason is because another dependency's constraints are blocking updates for the `ruby-lsp` gem. You should look for other gems that depend on `rbs`, `prism` or `sorbet-runtime` and try to ensure they are up to date.

Not being able to upgrade the language server is troublesome and the optimal solution is to figure out why the upgrades are failing and then fix it. Read the rest of the issue for more context.

### How to avoid the issues temporarily

Turn off ERB support until you are able to upgrade the `ruby-lsp` gem.

```jsonc
// Your VS Code settings
{
  "rubyLsp.erbSupport": false
}
```
**Note**: this setting is a temporary escape hatch. We will not support old versions of the language server indefinitely. After a few months of allowing users to turn off ERB support, we will remove the setting and go back to turning on support for all users.

**Comment by remy727** [comment]
Updated to the latest version and added `"rubyLsp.erbSupport": false` but I still have the issue.

**Comment by vinistock** [comment]
We haven't cut a release that includes the PR adding the setting yet.

**Comment by Adeynack** [comment]
I removed `Gemfile.lock` and ran `bundle install` again. Only then did it disappear for me.

**Comment by andreynering** [comment]
`"rubyLsp.erbSupport": false,` worked for me!

**Comment by goalaleo** [comment]
Updating `ruby-lsp` gem to `0.17.17` and `prism` to `1.0.0` worked for me

**Comment by pruzicka** [comment]
I had the same issue. Removed rufo & solargraph from Gemfile, bundle install. Problem solved, thanks for hints

**Comment by philsmy** [comment]
> I had the same issue. Removed rufo & solargraph from Gemfile, bundle install. Problem solved, thanks for hints

Yes, that was all I needed to do. I didn't have `rufo` in my Gemfile, but I did have Solargraph. I removed it, did `bundle i` and now that formatting madness has gone away!

**Comment by ptenteromano** [comment]
Adding the following to my settings.json worked for me:
```json
"[erb]": {
    "editor.defaultFormatter": "Shopify.ruby-lsp"
}
```
